File: seoBridge.js
'use strict';
/**
 * SEO Bridge — connect any Node.js service to the SEO Traffic Engine.
 * Usage: const SEOBridge = require('./seoBridge');
 *        const bridge = new SEOBridge({ projectName: 'my-app', keywords: ['shopify'] });
 *        bridge.startBackgroundSync();
 */
const https = require('https');
const http = require('http');

const SEO_ENGINE_URL = process.env.SEO_ENGINE_URL || 'https://seo-traffic-engine-production.up.railway.app';

class SEOBridge {
    constructor({ projectName, keywords = [], intervalHours = 6 } = {}) {
        this.projectName = projectName;
        this.keywords = keywords;
        this.interval = intervalHours * 3600 * 1000;
        this._timer = null;
    }

    async _fetch(method, path, body = null) {
        const url = new URL(path, SEO_ENGINE_URL);
        const isHttps = url.protocol === 'https:';
        const lib = isHttps ? https : http;
        return new Promise((resolve, reject) => {
            const opts = {
                hostname: url.hostname,
                port: url.port || (isHttps ? 443 : 80),
                path: url.pathname + url.search,
                method,
                headers: { 'Content-Type': 'application/json' },
                timeout: 10000,
            };
            const payload = body ? JSON.stringify(body) : null;
            if (payload) opts.headers['Content-Length'] = Buffer.byteLength(payload);
            const req = lib.request(opts, (res) => {
                let data = '';
                res.on('data', c => data += c);
                res.on('end', () => {
                    try { resolve(JSON.parse(data)); }
                    catch { resolve({}); }
                });
            });
            req.on('error', reject);
            req.on('timeout', () => { req.destroy(); reject(new Error('timeout')); });
            if (payload) req.write(payload);
            req.end();
        });
    }

    async pushKeywords(keywords) {
        const kws = keywords || this.keywords;
        const results = [];
        for (const kw of kws) {
            try {
                await this._fetch('POST', '/api/trigger/articles', { keyword: kw, project: this.projectName });
                results.push({ keyword: kw, ok: true });
            } catch (e) {
                results.push({ keyword: kw, ok: false, error: e.message });
            }
        }
        return results;
    }

    async getProducts(keyword, source = 'all') {
        try {
            const path = `/api/products?keyword=${encodeURIComponent(keyword)}&source=${source}`;
            const data = await this._fetch('GET', path);
            return data.products || [];
        } catch (e) { return []; }
    }

    async getRecommendations(topic) {
        try {
            return await this._fetch('POST', '/api/recommend', { topic, limit: 5 });
        } catch (e) { return { topic, amazon: [], ebay: [], total: 0 }; }
    }

    async getStats() {
        try {
            return await this._fetch('GET', '/stats');
        } catch (e) { return {}; }
    }

    startBackgroundSync() {
        const sync = async () => {
            try {
                await this.pushKeywords();
                console.log(`[SEOBridge] Synced ${this.keywords.length} keywords for ${this.projectName}`);
            } catch (e) {
                console.warn('[SEOBridge] Sync error:', e.message);
            }
        };
        // First sync after 2 minutes
        setTimeout(sync, 120000);
        this._timer = setInterval(sync, this.interval);
        return this;
    }

    // Add Express.js routes
    addExpressRoutes(app) {
        app.get('/api/seo', async (req, res) => {
            const stats = await this.getStats();
            res.json({
                project: this.projectName,
                seo_engine: SEO_ENGINE_URL,
                articles_tracked: (stats.articles || []).length,
                keywords: this.keywords.slice(0, 10),
                top_articles: (stats.articles || []).slice(0, 5),
            });
        });
        app.get('/api/seo/products', async (req, res) => {
            const keyword = req.query.keyword || this.keywords[0] || 'shopify';
            const products = await this.getProducts(keyword);
            res.json({ keyword, products });
        });
    }
}

module.exports = SEOBridge;

File: seo_bridge.py
"""
SEO Bridge — connect any service to the SEO Traffic Engine.
Usage:
    from seo_bridge import SEOBridge
    bridge = SEOBridge(project_name="my-project", keywords=["shopify", "seo"])
    asyncio.create_task(bridge.start_background_sync())
"""
import asyncio
import logging
import os

import aiohttp

log = logging.getLogger("seo_bridge")
SEO_ENGINE_URL = os.getenv("SEO_ENGINE_URL", "https://seo-traffic-engine-production.up.railway.app")


class SEOBridge:
    def __init__(self, project_name: str, keywords: list[str], interval_hours: int = 6):
        self.project_name = project_name
        self.keywords = keywords
        self.interval = interval_hours * 3600
        self._last_sync = 0

    async def push_keywords(self, keywords: list[str] | None = None) -> bool:
        kws = keywords or self.keywords
        try:
            async with aiohttp.ClientSession() as s:
                for kw in kws:
                    payload = {
                        "title": kw,
                        "url": os.getenv("APP_URL", ""),
                        "keyword": kw,
                        "excerpt": f"Article about {kw} for {self.project_name}",
                    }
                    async with s.post(
                        f"{SEO_ENGINE_URL}/api/trigger/articles",
                        json=payload,
                        timeout=aiohttp.ClientTimeout(total=8),
                    ) as r:
                        if r.status == 200:
                            log.info(f"SEO: triggered article for '{kw}'")
            return True
        except Exception as e:
            log.warning(f"SEO push failed: {e}")
            return False

    async def get_products(self, keyword: str, source: str = "all") -> list[dict]:
        try:
            async with aiohttp.ClientSession() as s:
                async with s.get(
                    f"{SEO_ENGINE_URL}/api/products",
                    params={"keyword": keyword, "source": source},
                    timeout=aiohttp.ClientTimeout(total=10),
                ) as r:
                    if r.status == 200:
                        d = await r.json()
                        return d.get("products", [])
        except Exception as e:
            log.warning(f"SEO products failed: {e}")
        return []

    async def get_recommendations(self, topic: str) -> dict:
        try:
            async with aiohttp.ClientSession() as s:
                async with s.post(
                    f"{SEO_ENGINE_URL}/api/recommend",
                    json={"topic": topic, "limit": 5},
                    timeout=aiohttp.ClientTimeout(total=10),
                ) as r:
                    if r.status == 200:
                        return await r.json()
        except Exception as e:
            log.warning(f"SEO recommend failed: {e}")
        return {"topic": topic, "amazon": [], "ebay": [], "total": 0}

    async def get_articles(self) -> list[dict]:
        try:
            async with aiohttp.ClientSession() as s:
                async with s.get(
                    f"{SEO_ENGINE_URL}/stats",
                    timeout=aiohttp.ClientTimeout(total=8),
                ) as r:
                    if r.status == 200:
                        d = await r.json()
                        return d.get("articles", [])
        except Exception as e:
            log.warning(f"SEO stats failed: {e}")
        return []

    async def start_background_sync(self):
        import time

        await asyncio.sleep(60)
        while True:
            try:
                now = time.time()
                if now - self._last_sync >= self.interval:
                    await self.push_keywords()
                    self._last_sync = now
                    log.info(f"SEO sync done for {self.project_name}")
            except Exception as e:
                log.warning(f"SEO sync error: {e}")
            await asyncio.sleep(3600)


def add_seo_routes(router, bridge: "SEOBridge"):
    """Add /api/seo and /api/seo/products routes to an aiohttp router."""
    from aiohttp import web

    async def handle_seo(request):
        articles = await bridge.get_articles()
        return web.json_response(
            {
                "project": bridge.project_name,
                "seo_engine": SEO_ENGINE_URL,
                "articles_tracked": len(articles),
                "keywords": bridge.keywords[:10],
                "top_articles": articles[:5],
            }
        )

    async def handle_seo_products(request):
        keyword = request.rel_url.query.get(
            "keyword", bridge.keywords[0] if bridge.keywords else "shopify"
        )
        products = await bridge.get_products(keyword)
        return web.json_response({"keyword": keyword, "products": products})

    router.add_get("/api/seo", handle_seo)
    router.add_get("/api/seo/products", handle_seo_products)
